File: agents/windows/src/packet_parser.cpp
#include "edr_agent/packet_parser.hpp"

#include <algorithm>
#include <set>
#include <sstream>

namespace edr {
namespace {

std::optional<std::uint16_t> u16(std::span<const std::uint8_t> bytes, std::size_t offset) {
    if (offset + 1 >= bytes.size()) return std::nullopt;
    return static_cast<std::uint16_t>((bytes[offset] << 8) | bytes[offset + 1]);
}

std::string tls_version(std::uint16_t value) {
    switch (value) {
        case 0x0301: return "TLS1.0";
        case 0x0302: return "TLS1.1";
        case 0x0303: return "TLS1.2";
        case 0x0304: return "TLS1.3";
        default: return "UNKNOWN";
    }
}

}  // namespace

std::vector<Event> PacketParser::parse_ethernet(std::span<const std::uint8_t> packet) {
    if (packet.size() < 34 || !u16(packet, 12) || *u16(packet, 12) != 0x0800) return {};
    const std::size_t ip = 14;
    const std::size_t ip_header = static_cast<std::size_t>(packet[ip] & 0x0f) * 4;
    if (ip_header < 20 || packet.size() < ip + ip_header) { ++parse_errors_; return {}; }
    if (packet[ip + 9] == 17) return parse_udp(packet, ip + ip_header);
    if (packet[ip + 9] == 6) return parse_tcp(packet, ip + ip_header);
    return {};
}

std::vector<Event> PacketParser::parse_udp(std::span<const std::uint8_t> packet, std::size_t offset) {
    if (packet.size() < offset + 20) { ++parse_errors_; return {}; }
    const auto source = u16(packet, offset), destination = u16(packet, offset + 2);
    if (!source || !destination || (*source != 53 && *destination != 53)) return {};
    const auto dns = packet.subspan(offset + 8);
    std::size_t cursor = 12;
    std::vector<std::string> labels;
    while (cursor < dns.size()) {
        const std::size_t length = dns[cursor++];
        if (length == 0) break;
        if (length > 63 || cursor + length > dns.size()) { ++parse_errors_; return {}; }
        labels.emplace_back(reinterpret_cast<const char*>(dns.data() + cursor), length);
        cursor += length;
    }
    const auto query_type = u16(dns, cursor);
    if (labels.empty() || !query_type) { ++parse_errors_; return {}; }
    std::string query;
    for (const auto& label : labels) { if (!query.empty()) query += '.'; query += label; }
    const std::map<int, std::string> types{{1, "A"}, {5, "CNAME"}, {12, "PTR"}, {28, "AAAA"}};
    const auto found = types.find(*query_type);
    return {{uuid_string(), "DNS_QUERY", utc_now(), {
        {"query", query}, {"recordType", found == types.end() ? std::to_string(*query_type) : found->second}
    }}};
}

std::vector<Event> PacketParser::parse_tcp(std::span<const std::uint8_t> packet, std::size_t offset) {
    if (packet.size() < offset + 20) { ++parse_errors_; return {}; }
    const std::size_t tcp_header = static_cast<std::size_t>((packet[offset + 12] >> 4) & 0x0f) * 4;
    if (tcp_header < 20 || packet.size() <= offset + tcp_header) return {};
    const auto payload = packet.subspan(offset + tcp_header);
    const std::string text(reinterpret_cast<const char*>(payload.data()), std::min<std::size_t>(payload.size(), 16384));
    const auto line_end = text.find("\r\n");
    const std::string first = text.substr(0, line_end);
    const std::set<std::string> methods{"GET", "POST", "PUT", "PATCH", "DELETE", "HEAD", "OPTIONS"};
    const auto space = first.find(' ');
    if (space != std::string::npos && methods.contains(first.substr(0, space))) {
        const auto next = first.find(' ', space + 1);
        std::map<std::string, JsonValue> result{
            {"l7Protocol", std::string("HTTP")}, {"httpMethod", first.substr(0, space)},
            {"url", sanitize_url(first.substr(space + 1, next - space - 1))}
        };
        std::istringstream lines(text);
        std::string line;
        while (std::getline(lines, line)) {
            if (!line.empty() && line.back() == '\r') line.pop_back();
            if (line.rfind("Host:", 0) == 0) result["httpHost"] = line.substr(5 + (line.size() > 5 && line[5] == ' '));
            else if (line.rfind("User-Agent:", 0) == 0) result["httpUserAgent"] = line.substr(11 + (line.size() > 11 && line[11] == ' '));
        }
        return {{uuid_string(), "L7_EVENT", utc_now(), std::move(result)}};
    }
    if (payload.size() < 9 || payload[0] != 22 || payload[5] != 1) return {};
    std::size_t cursor = 9;
    if (cursor + 34 > payload.size()) { ++parse_errors_; return {}; }
    const auto version = u16(payload, cursor).value_or(0x0303);
    cursor += 34;
    if (cursor >= payload.size()) return {};
    cursor += 1 + payload[cursor];
    const auto cipher_length = u16(payload, cursor); if (!cipher_length) return {};
    cursor += 2 + *cipher_length; if (cursor >= payload.size()) return {};
    cursor += 1 + payload[cursor];
    const auto extension_length = u16(payload, cursor); if (!extension_length) return {};
    cursor += 2;
    const auto end = std::min(payload.size(), cursor + *extension_length);
    std::optional<std::string> sni;
    while (cursor + 4 <= end) {
        const auto type = u16(payload, cursor), length = u16(payload, cursor + 2); cursor += 4;
        if (!type || !length || cursor + *length > end) { ++parse_errors_; break; }
        if (*type == 0 && *length >= 5) {
            const auto name_length = u16(payload, cursor + 3);
            if (name_length && cursor + 5 + *name_length <= end) {
                sni = std::string(reinterpret_cast<const char*>(payload.data() + cursor + 5), *name_length);
            }
        }
        cursor += *length;
    }
    std::map<std::string, JsonValue> result{{"l7Protocol", std::string("TLS")}, {"tlsVersion", tls_version(version)}};
    if (sni) result["tlsSni"] = *sni;
    return {{uuid_string(), "L7_EVENT", utc_now(), std::move(result)}};
}

}  // namespace edr

File: agents/macos/Sources/EDRAgentCore/PacketParser.swift
import Foundation

private extension Data {
    func u16(_ offset: Int) -> Int? {
        guard offset >= 0, offset + 1 < count else { return nil }
        return (Int(self[offset]) << 8) | Int(self[offset + 1])
    }

    func u32(_ offset: Int, littleEndian: Bool) -> Int? {
        guard offset >= 0, offset + 3 < count else { return nil }
        let bytes = (0..<4).map { Int(self[offset + $0]) }
        return littleEndian
            ? bytes[0] | (bytes[1] << 8) | (bytes[2] << 16) | (bytes[3] << 24)
            : (bytes[0] << 24) | (bytes[1] << 16) | (bytes[2] << 8) | bytes[3]
    }
}

public final class PacketStreamParser: @unchecked Sendable {
    private var buffer = Data()
    private var headerParsed = false
    private var littleEndian = true
    public private(set) var parseErrorCount = 0

    public init() {}

    public func append(_ data: Data) -> [TelemetryEvent] {
        buffer.append(data)
        var events: [TelemetryEvent] = []
        if !headerParsed {
            guard buffer.count >= 24 else { return [] }
            let magic = Array(buffer.prefix(4))
            if magic == [0xd4, 0xc3, 0xb2, 0xa1] { littleEndian = true }
            else if magic == [0xa1, 0xb2, 0xc3, 0xd4] { littleEndian = false }
            else { parseErrorCount += 1; buffer.removeAll(); return [] }
            buffer.removeFirst(24)
            headerParsed = true
        }
        while buffer.count >= 16 {
            guard let includedLength = buffer.u32(8, littleEndian: littleEndian), includedLength >= 0, includedLength <= 1_048_576 else {
                parseErrorCount += 1; buffer.removeAll(); break
            }
            guard buffer.count >= 16 + includedLength else { break }
            let packet = Data(buffer[16..<(16 + includedLength)])
            buffer.removeFirst(16 + includedLength)
            events.append(contentsOf: parseEthernet(packet))
        }
        return events
    }

    public func parseEthernet(_ packet: Data) -> [TelemetryEvent] {
        guard packet.count >= 34, packet.u16(12) == 0x0800 else { return [] }
        let ipOffset = 14
        let headerLength = Int(packet[ipOffset] & 0x0f) * 4
        guard headerLength >= 20, packet.count >= ipOffset + headerLength else { parseErrorCount += 1; return [] }
        let protocolNumber = packet[ipOffset + 9]
        let transportOffset = ipOffset + headerLength
        if protocolNumber == 17 { return parseUDP(packet, offset: transportOffset) }
        if protocolNumber == 6 { return parseTCP(packet, offset: transportOffset) }
        return []
    }

    private func parseUDP(_ packet: Data, offset: Int) -> [TelemetryEvent] {
        guard packet.count >= offset + 8, let sourcePort = packet.u16(offset), let destinationPort = packet.u16(offset + 2) else {
            parseErrorCount += 1; return []
        }
        guard sourcePort == 53 || destinationPort == 53 else { return [] }
        let dns = Data(packet[(offset + 8)...])
        guard dns.count >= 12 else { parseErrorCount += 1; return [] }
        var cursor = 12
        var labels: [String] = []
        while cursor < dns.count {
            let length = Int(dns[cursor]); cursor += 1
            if length == 0 { break }
            guard length <= 63, cursor + length <= dns.count else { parseErrorCount += 1; return [] }
            labels.append(String(decoding: dns[cursor..<(cursor + length)], as: UTF8.self))
            cursor += length
        }
        guard !labels.isEmpty, let queryType = dns.u16(cursor) else { parseErrorCount += 1; return [] }
        let recordType = [1: "A", 2: "NS", 5: "CNAME", 12: "PTR", 15: "MX", 16: "TXT", 28: "AAAA"][queryType] ?? String(queryType)
        let responseCode = Int(dns[3] & 0x0f)
        var payload: [String: JSONValue] = ["query": .string(labels.joined(separator: ".")), "recordType": .string(recordType)]
        if dns[2] & 0x80 != 0 { payload["responseCode"] = .string(responseCode == 0 ? "NOERROR" : String(responseCode)) }
        return [TelemetryEvent(eventType: "DNS_QUERY", payload: payload)]
    }

    private func parseTCP(_ packet: Data, offset: Int) -> [TelemetryEvent] {
        guard packet.count >= offset + 20 else { parseErrorCount += 1; return [] }
        let headerLength = Int((packet[offset + 12] >> 4) & 0x0f) * 4
        guard headerLength >= 20, packet.count > offset + headerLength else { return [] }
        let payload = Data(packet[(offset + headerLength)...])
        if let http = parseHTTP(payload) { return [http] }
        if let tls = parseTLSClientHello(payload) { return [tls] }
        return []
    }

    private func parseHTTP(_ payload: Data) -> TelemetryEvent? {
        guard let text = String(data: payload.prefix(16_384), encoding: .utf8) else { return nil }
        let lines = text.components(separatedBy: "\r\n")
        guard let first = lines.first else { return nil }
        var result: [String: JSONValue] = ["l7Protocol": .string("HTTP")]
        if first.hasPrefix("HTTP/") {
            let fields = first.split(separator: " ")
            if fields.count >= 2, let status = Int(fields[1]) { result["httpStatusCode"] = .integer(status) }
        } else {
            let fields = first.split(separator: " ")
            let methods = Set(["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD", "OPTIONS"])
            guard fields.count >= 2, methods.contains(String(fields[0])) else { return nil }
            result["httpMethod"] = .string(String(fields[0]))
            result["url"] = .string(sanitizeURL(String(fields[1])))
        }
        for line in lines.dropFirst() {
            let lower = line.lowercased()
            if lower.hasPrefix("host:") { result["httpHost"] = .string(line.dropFirst(5).trimmingCharacters(in: .whitespaces)) }
            else if lower.hasPrefix("user-agent:") { result["httpUserAgent"] = .string(line.dropFirst(11).trimmingCharacters(in: .whitespaces)) }
        }
        return TelemetryEvent(eventType: "L7_EVENT", payload: result)
    }

    private func parseTLSClientHello(_ payload: Data) -> TelemetryEvent? {
        guard payload.count >= 9, payload[0] == 22, payload[5] == 1 else { return nil }
        var cursor = 9
        guard cursor + 34 <= payload.count else { parseErrorCount += 1; return nil }
        let legacyVersion = payload.u16(cursor) ?? 0x0303
        cursor += 34
        guard cursor < payload.count else { return nil }
        let sessionLength = Int(payload[cursor]); cursor += 1 + sessionLength
        guard let cipherLength = payload.u16(cursor) else { return nil }
        cursor += 2 + cipherLength
        guard cursor < payload.count else { return nil }
        let compressionLength = Int(payload[cursor]); cursor += 1 + compressionLength
        guard let extensionsLength = payload.u16(cursor) else { return nil }
        cursor += 2
        let extensionsEnd = min(payload.count, cursor + extensionsLength)
        var sni: String?
        while cursor + 4 <= extensionsEnd {
            guard let type = payload.u16(cursor), let length = payload.u16(cursor + 2) else { break }
            cursor += 4
            guard cursor + length <= extensionsEnd else { parseErrorCount += 1; break }
            if type == 0, length >= 5, let nameLength = payload.u16(cursor + 3), cursor + 5 + nameLength <= extensionsEnd {
                sni = String(decoding: payload[(cursor + 5)..<(cursor + 5 + nameLength)], as: UTF8.self)
            }
            cursor += length
        }
        var result: [String: JSONValue] = ["l7Protocol": .string("TLS"), "tlsVersion": .string(tlsVersion(legacyVersion))]
        if let sni, !sni.isEmpty { result["tlsSni"] = .string(sni) }
        return TelemetryEvent(eventType: "L7_EVENT", payload: result)
    }

    private func sanitizeURL(_ value: String) -> String {
        String(value.prefix { $0 != "?" && $0 != "#" })
    }

    private func tlsVersion(_ value: Int) -> String {
        switch value { case 0x0301: "TLS1.0"; case 0x0302: "TLS1.1"; case 0x0303: "TLS1.2"; case 0x0304: "TLS1.3"; default: String(format: "0x%04x", value) }
    }
}

public struct PacketCollection: Sendable {
    public let events: [TelemetryEvent]
    public let health: [SensorSnapshot]
}

public enum TcpdumpCollector {
    public static let executable = "/usr/sbin/tcpdump"

    public static func collect(interface: String, seconds: TimeInterval) -> PacketCollection {
        let process = Process(), output = Pipe(), error = Pipe()
        process.executableURL = URL(fileURLWithPath: executable)
        process.arguments = ["-i", interface, "-U", "-n", "-s", "0", "-w", "-"]
        process.standardOutput = output
        process.standardError = error
        do { try process.run() } catch { return degraded(parseErrors: 0) }
        Thread.sleep(forTimeInterval: max(0.2, seconds))
        if process.isRunning { process.interrupt() }
        let bytes = output.fileHandleForReading.readDataToEndOfFile()
        let diagnostic = String(decoding: error.fileHandleForReading.readDataToEndOfFile(), as: UTF8.self)
        process.waitUntilExit()
        let denied = diagnostic.localizedCaseInsensitiveContains("permission denied") || diagnostic.localizedCaseInsensitiveContains("BIOCSETIF")
        guard !denied, process.terminationStatus == 0 || !bytes.isEmpty else { return degraded(parseErrors: 0) }
        let parser = PacketStreamParser()
        let events = parser.append(bytes)
        let drops = packetDrops(diagnostic)
        return PacketCollection(events: events, health: [
            SensorSnapshot(sensor: "PACKET_METADATA", status: "HEALTHY", provider: "TCPDUMP", packetDropCount: drops),
            SensorSnapshot(sensor: "L7", status: "HEALTHY", parseErrorCount: parser.parseErrorCount),
        ])
    }

    private static func degraded(parseErrors: Int) -> PacketCollection {
        PacketCollection(events: [], health: [
            SensorSnapshot(sensor: "PACKET_METADATA", status: "DEGRADED", provider: "TCPDUMP", packetDropCount: 0),
            SensorSnapshot(sensor: "L7", status: "DEGRADED", parseErrorCount: parseErrors),
        ])
    }

    private static func packetDrops(_ diagnostic: String) -> Int {
        let expression = try? NSRegularExpression(pattern: "([0-9]+) packets dropped by kernel")
        let range = NSRange(diagnostic.startIndex..., in: diagnostic)
        guard let match = expression?.firstMatch(in: diagnostic, range: range),
              let valueRange = Range(match.range(at: 1), in: diagnostic) else { return 0 }
        return Int(diagnostic[valueRange]) ?? 0
    }
}
